### PR TITLE
Unify logic to checkout a pull request

### LINKED ISSUE #9439: Unify logic to checkout PRs
Currently, there are two ways to checkout a PR:

1. From the "Pull Requests" list tab that appears on the GitHub Desktop app.
2. From github.com when navigating to a Pull Request page.

Each of these entry points has its own logic, which causes a few inconsistencies/bugs mostly on the checkout from dotcom flow.

Unifying this logic will make it much easier to fix bugs and keep the behaviour more consistent between the two flows.

(note that these bugs can be reproduced in current production).

## 1. PR branches created in dotcom don't get associated to the PR

How to reproduce:

* Clone a repository that has PRs (e.g `desktop/desktop`) from Desktop.
* Open a the dotcom page of a PR that was created from a fork (e.g `https://github.com/desktop/desktop/pull/9430`).
* Click on the `open this in GitHub Desktop` link.

**Expected behaviour:** The PR branch gets checked out and it gets linked to the PR in the desktop UI:

Screenshot 2020-04-02 at 17 20 49

**Actual behaviour:** The PR branch does not get linked to the PR:

Screenshot 2020-04-02 at 17 19 58

## 2. Cannot checkout some PRs from dotcom

There's a quite rare situation that can cause an error popping out when trying to checkout some specific PRs from dotcom (@tierninho ran into it in https://github.com/desktop/desktop/pull/9340#issuecomment-604164125).

Steps to reproduce: 

* Clone `kuychaco/urban-potato` from Desktop.
* Open `https://github.com/not-kuychaco/urban-potato/pull/3` and click on the `open this in GitHub Desktop` link.

**Expected behaviour:** The PR branch gets checked out.

**Actual behaviour:** The following error appears:

Screenshot 2020-04-02 at 17 49 38

## 3. Wrong name of PR branches created from Desktop

When checking out a PR created on an upstream repository from Desktop, in some situations (when the PR was created from a branch in the upstream repository) the local branch will have an incorrect name and the `upstream` remote will get duplicated.

(this issue is only reproducible with the changes in https://github.com/desktop/desktop/pull/9340, since before that the upstream PRs were not shown).

Steps to reproduce:

* Clone `kuychaco/urban-potato` from Desktop.
* Go to the list of PRs inside Desktop and click on the "Update readme.md" PR (PR number 3).

**Expected behaviour**: A local branch called `kuychaco-patch-1` is created tracking `upstream/kuychaco-patch-1` (the branch of the PR).

**Actual behaviour**: A new remote `github-desktop-not-kuychaco` is added to the repo with the same value as the `upstream` remote and a local branch called `pr/3` is created tracking `github-desktop-not-kuychaco/kuychaco-patch-1`.

(this is not a big issue, but it's inconsistent with what happens when checking out the same PR from dotcom, see https://github.com/desktop/desktop/pull/9340#issuecomment-603935866).

### LINKED ISSUE #9593: Possible error with remotes when you want to checkout upstream PRs on Desktop
### Describe the bug

When I'm working with a fork, in my case with Desktop, and want to checkout upstream PRs, github-desktop-desktop remote is created even if upstream remote exists.

### Version & OS

GitHub Desktop 5c0df17.
OS: Windows 10 

### Steps to reproduce the behavior

1. Select a repository fork
2. Checkout a PR (from upstream)
3. se github-desktop-*** remote created

### Expected behavior

No remote is created. Upstream remote used

### Actual behavior

github-desktop-*** (github-desktop-desktop in my case) remote created as duplicate of upstream.

### Screenshots

Add screenshots to help explain your problem, if applicable.

### Logs

Attach your logs by opening the `Help` menu and selecting `Show Logs...`, if applicable.

### Additional context

Think the problem is here:
https://github.com/desktop/desktop/blob/5c0df1747a85e0c3f383bb82866faf9d7ad2b5c5/app/src/lib/stores/app-store.ts#L5329-L5330
Should be:
```ts
const isRefInThisRepo = gitHubRepository.fork ?
  head.gitHubRepository.cloneURL === gitHubRepository.parent!.cloneURL :
  head.gitHubRepository.cloneURL === gitHubRepository.cloneURL
```
This works in my case

### PR DESCRIPTION
**Pull request by rafeca**
Closes #9439
Closes #9593

This PR is based off https://github.com/desktop/desktop/pull/9340

## Description

This PR contains a small refactor to unify the logic to checkout a PR from GitHub.

Currently, there are two ways to checkout a PR:

1. From the "Pull Requests" list tab that appears on the GitHub Desktop app.
    Screenshot 2020-03-27 at 15 42 47
2. From github.com when navigating to a Pull Request page:
    Screenshot 2020-03-27 at 15 43 15

Each of these entry points has its own logic, which causes some inconsistencies between them (as we found in https://github.com/desktop/desktop/pull/9340#issuecomment-603953203).

## Test cases

I've tried to follow the test cases that were written in https://github.com/desktop/desktop/pull/9138, but couldn't check every single one of them.

## Release notes

Notes: no-notes

### CODE REVIEW

File: app/src/lib/stores/app-store.ts [hunk 1]
```diff
@@ -5271,18 +5272,47 @@ export class AppStore extends TypedBaseStore<IAppState> {
...
+        return
+      }
+
+      prNumber = pullRequest.number
+      ownerLogin = pullRequest.user.login
+      head = {
+        cloneURL: pullRequest.head.repo.clone_url,
+        ref: pullRequest.head.ref,
+      }
+    }
```
Comment (rafeca) [inline app/src/lib/stores/app-store.ts]: Not particularly proud of this piece of logic, but it allows the `_checkoutPullRequest()` method to accept both a `PullRequest` object and an `IAPIPullRequest` one.

I can also move this logic into a separate method to try to keep this method cleaner.

File: app/src/ui/dispatcher/dispatcher.ts [hunk 2]
```diff
@@ -1400,99 +1404,106 @@ export class Dispatcher {
...
+
+    if (headUrl === null || baseUrl === null) {
+      return null
+    }
+
+    for (const repository of repositories) {
+      if (this.doesRepositoryMatchUrl(repository, headUrl)) {
+        return repository
+      }
+    }
```
Comment (rafeca) [inline app/src/ui/dispatcher/dispatcher.ts]: I wanted to use `repositories.find()` here but TypeScript's `Array.prototype.find` is not capable to infer that the found element can only be of type `Repository | null`.

**Comment by tierninho** [comment]
Ran through almost all the tests here https://github.com/desktop/desktop/pull/9138#issuecomment-592714953 and only one stood out -- the rest passed.

>  PR in upstream with upstream as source —> opens fork in Desktop and source branch is checked out (not-kuychaco/urban-potato#3)

To set up:
- Fork https://github.com/not-kuychaco/urban-potato, stay on master branch
- Go to any other repo on Desktop
- Go to https://github.com/not-kuychaco/urban-potato/pull/3 and click Open with Desktop link. 

Expected: branch opens in forked repo
Actual: the Clone Repository modal opens

**Comment by rafeca** [comment]
>Ran through almost all the tests here #9138 (comment) and only one stood out -- the rest passed.

Thanks! ✨✨

I fixed the issue in the last commit. This was caused because the URL that's opened in that case is `x-github-client://openRepo/https://github.com/not-kuychaco/urban-potato?branch=kuychaco-patch-1`, when it should be something like `x-github-client://openRepo/https://github.com/not-kuychaco/urban-potato?pr=3`. I fixed the issue in Desktop, but I'll try to change the logic in dotcom to make the URLs from all PRs consistent.

**Comment by niik** [comment]
> Hey, I was looking to test some of the scenarios here, but is there a merge conflict that needs resolving?

@tierninho I think this is because @rafeca force-pushed the branch at one point

image

Unfortunately for you it means it's gonna be a bit trickier to get up to date. Either you delete the branch locally and check it out again or you can run the following command in your terminal

```git reset --hard  origin/techdebt/unify-checkout-pr-logic```

**Comment by rafeca** [comment]
> @tierninho I think this is because @rafeca force-pushed the branch at one point

Ups, sorry about that @tierninho ! and thanks @niik for figuring it out!

**Review by niik — APPROVED** [review]
Great work cleaning this up @rafeca!

**Pull request merged**